### PR TITLE
pkg/torcx: introduce an image manifest for propagation phase

### PR DESCRIPTION
**Pull request by lucab**
This introduces a new `.torcx/manifest.json` to specify which assets contained in the image needs to be propagated to the host system. It currently supports propagating binaries (to `/var/run/torcx/bin`) and systemd service units (to runtime `/run/systemd/system`).

Closes: #29

### CODE REVIEW

**Comment by lucab** [comment]
Updated, @squeed PTAL.

File: examples/image-manifest.json [hunk 1]
```diff
@@ -0,0 +1,12 @@
+{
+  "kind": "image-manifest-v0",
+  "value": {
+      "bin": [
+          "/usr/sbin/cupsd"
+      ],
+      "service": [
+          "/lib/systemd/system/cups.service",
+          "/lib/systemd/system/printer.target.wants/"
```
Comment (squeed) [inline examples/image-manifest.json]: Does this imply the whole directory is copied?

File: pkg/torcx/propagate.go [hunk 2]
```diff
@@ -15,94 +15,197 @@
...
+				return nil
 			}
-			return os.Symlink(linkDest, dstPath)
-		} else if srcStat.Mode().IsRegular() {
-			// Normal files: symlink in to the systemd volatile directory
-			logrus.WithFields(logrus.Fields{
-				"src": srcPath,
-				"dst": dstPath,
-			}).Debug("installing unit file")
-			return os.Symlink(srcPath, dstPath)
```
Comment (squeed) [inline pkg/torcx/propagate.go]: Hmm, am I missing something? How do the .wants symlinks get created?

File: examples/image-manifest.json (hunk 1 above)
Comment (squeed) [inline examples/image-manifest.json]: Nevermind, I see it!

File: pkg/torcx/propagate.go [hunk 3]
```diff
@@ -15,94 +15,197 @@
...
+		if err := flattenBinAsset(applyCfg, applyCfg.RunBinDir(), path); err != nil {
 			return err
 		}
+	}
+	return nil
+}
+
+// flattenBinAssets propagates a single binary or a directory of binaries,
+// flattening all intermediate directories.
+func flattenBinAsset(applyCfg *ApplyConfig, binDir string, asset string) error {
```
Comment (squeed) [inline pkg/torcx/propagate.go]: Bikeshedding: The name "flatten" is a bit confusing to me. I associate it with collection code, and don't expect it to have side-effects. Maybe call it `propagateBinAsset`?

**Comment by squeed** [comment]
Seems reasonable. For docker, we'll need the ability to propagate `systemd-networkd` files as well, if you want to think about that.

File: examples/image-manifest.json (hunk 1 above)
Comment (lucab) [inline examples/image-manifest.json]: Just for reference: yes, the whole directory is copied. If it already exists, its content is merged in the existing one without overwriting existing assets.

File: pkg/torcx/propagate.go (hunk 3 above)
Comment (lucab) [inline pkg/torcx/propagate.go]: `propagateBins` is already the parent caller of this, while the purpose of this is to flatten any specified assets directory and just symlink every single binary in it. Perhaps `symlinkBinAsset`?

File: pkg/torcx/propagate.go (hunk 2 above)
Comment (lucab) [inline pkg/torcx/propagate.go]: No, you are right I forgot a case here.

**Comment by lucab** [comment]
Second iteration, PTAL:
 * renamed `flatten*` to `symlink*`
 * fixed symlinked-units propagation

I'm keeping network and udev assets out of this on purpose, I'll attack them in a later after diggin a bit in networkd and udevd setup in CL.

**Comment by squeed** [comment]
LGTM, awesome!

**Pull request merged**